tests/test_defaults.py

```diff
@@ -84,3 +84,43 @@ def foo(name):
 
     result = runner.invoke(cli, ["foo", "--help"], default_map={"foo": {"name": True}})
     assert "default: name" in result.output
+
+
+def test_flag_default_bool(runner):
+    """test flag with default bool"""
+
+    @click.command()
+    @click.option("--foo", is_flag=True, default=False, type=click.BOOL)
+    @click.option("--bar", is_flag=True, default=False, type=bool)
+    @click.option("--baz", is_flag=True, default=True, type=click.BOOL)
+    @click.option("--qux", is_flag=True, default=True, type=bool)
+    def cli(foo, bar, baz, qux):
+        assert isinstance(foo, bool)
+        assert isinstance(bar, bool)
+        assert isinstance(baz, bool)
+        assert isinstance(qux, bool)
+        click.echo(f"foo: {foo}, bar: {bar}, baz: {baz}, qux: {qux}")
+
+    result = runner.invoke(cli, [])
```

Comment:
Not sure if this is the correct place for this test but I think it's required as this case was not covered introducing this breaking change.